Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP1Connections.swift

```diff
@@ -0,0 +1,602 @@
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+        }
+
+        /// The connection failed to start
+        mutating func failedToStart() {
+            switch self.state {
+            case .starting:
+                self.state = .backingOff
+            case .backingOff, .available, .leased, .closed:
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+        }
+
+        mutating func lease() -> Connection {
+            switch self.state {
+            case .available(let connection, since: _):
+                self.state = .leased(connection)
+                return connection
+            case .backingOff, .starting, .leased, .closed:
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+        }
+
+        mutating func release() {
+            switch self.state {
+            case .leased(let connection):
+                self.state = .available(connection, since: .now())
+            case .backingOff, .starting, .available, .closed:
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+        }
+
+        mutating func close() -> Connection {
+            switch self.state {
+            case .available(let connection, since: _):
+                self.state = .closed
+                return connection
+            case .backingOff, .starting, .leased, .closed:
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+        }
+
+        mutating func cancel() -> Connection {
+            switch self.state {
+            case .leased(let connection):
+                return connection
+            case .backingOff, .starting, .available, .closed:
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+        }
+
+        mutating func cleanup(_ context: inout CleanupContext) -> Bool {
+            switch self.state {
+            case .backingOff:
+                context.connectBackoff.append(self.connectionID)
+                return true
+            case .starting:
+                return false
+            case .available(let connection, since: _):
+                context.close.append(connection)
+                return true
+            case .leased(let connection):
+                context.cancel.append(connection)
+                return false
+            case .closed:
+                preconditionFailure("Unexpected state: Did not expect to have connections with this state in the state machine: \(self.state)")
+            }
+        }
+    }
+
+    /// A structure to hold the currently active HTTP/1.1 connections.
+    ///
+    /// The general purpose connection pool (pool for requests that don't have special `EventLoop`
+    /// requirements) will grow up until `maximumConcurrentConnections`. If requests have
+    /// special `EventLoop` requirements overflow connections might be opened.
```

Comment:
Just checking I understand: the size of the overflow isn't limited at all, we assume that the size of the overflow will be small.

Do we treat overflow connections differently? Once they are returned do we close them immediately or do we keep them around for some time?

Reply:
> the size of the overflow isn't limited at all, we assume that the size of the overflow will be small.
yes

> Once they are returned do we close them immediately or do we keep them around for some time?
I want to close overflow connections right away, once there isn't more work for them todo.

However with this basic setup changing this to whatever should be fairly simple :)

Reply:
It might be useful to document this to explain that while the pool can overflow we'll try to pull it back down to size when we can.

Reply:
> It might be useful to document this to explain that while the pool can overflow we'll try to pull it back down to size when we can.

This is finally the decision of the state machine. I don't want to document here, since this might get out of sync very fast.